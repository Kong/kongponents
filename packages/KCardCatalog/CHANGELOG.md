# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 0.1.1-alpha.0 (2021-11-16)

**Note:** Version bump only for package @kongponents/kcardcatalog
