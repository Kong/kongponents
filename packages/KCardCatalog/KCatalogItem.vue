<template>
  <router-link
    v-if="locationParam"
    :to="locationParam">
    <KCard
      :data-testid="`${item.name || item.title || item.id}`"
      class="grid-item d-flex flex-column overflow-hidden k-card-catalog-item"
      has-hover
      role="button"
      tabindex="0"
      @click.native="handleCardClick($event, item)"
    >
      <template #title>
        <slot
          :item="item"
          name="card-title"
        >
          {{ item.title }}
        </slot>
      </template>
      <template #body>
        <div
          :item="item"
          :class="{ 'multi-line-truncate': truncate }">
          <slot
            name="card-body"
          >
            {{ item.description }}
          </slot>
        </div>
      </template>
    </KCard>
  </router-link>

  <KCard
    v-else
    :data-testid="`${item.name || item.title || item.id}`"
    class="grid-item d-flex flex-column overflow-hidden k-card-catalog-item"
    has-hover
    role="button"
    tabindex="0"
    @click.native="handleCardClick($event, item)"
  >
    <template #title>
      <slot
        :item="item"
        name="card-title"
      >
        {{ item.title }}
      </slot>
    </template>
    <template #body>
      <div
        :item="item"
        :class="{ 'multi-line-truncate': truncate }">
        <slot
          name="card-body"
        >
          {{ item.description }}
        </slot>
      </div>
    </template>
  </KCard>
</template>

<script>
export default {
  name: 'KCatalogItem',
  props: {
    item: {
      type: Object,
      default: () => {},
      // Item must have a title
      validator: (items) => !items.length || items.some(i => i.hasOwnProperty('title'))
    },
    locationParam: {
      type: String,
      default: ''
    },
    truncate: {
      type: Boolean,
      default: true
    }
  },
  methods: {
    handleCardClick (evt, item) {
      this.$emit('clicked', {
        evt,
        item
      })
    }
  }
}
</script>

<style lang="scss">
.kong-card {
  --KCardBorderRadius: 5px;
  --KCardPaddingY: var(--spacing-lg);
  --KCardPaddingX: var(--spacing-md);

  display: flex;
  flex-direction: column;
  justify-content: stretch;

  .multi-line-truncate {
    --lh: 20px;
    --max-lines: 5;
    --max-lines-less: 4;
    line-height: var(--lh);
    position: relative;
    max-height: calc(var(--lh) * var(--max-lines));
    overflow: hidden;
    padding-right: 8px;
  }

  .multi-line-truncate::before {
    position: absolute;
    content: "...";
    top: calc(var(--lh) * var(--max-lines-less));
    right: 12px;
  }

  .k-card-body {
    flex-grow: 1;
    display: flex;
    flex-direction: column;

    .description {
      flex-grow: 1;
    }
  }

  &:hover {
    cursor: pointer;
    --KCardShadow: none;
    --KCardBorder: 1px solid var(--blue-300);
  }
}
</style>
