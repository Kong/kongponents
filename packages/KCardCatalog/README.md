# @kongponents/kcardcatalog

[![](https://img.shields.io/npm/v/@kongponents/kcardcatalog.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kcardcatalog)

```vue
<KCardCatalog :description="'hello world'">
  Hello from a slot
</KCardCatalog>
```
