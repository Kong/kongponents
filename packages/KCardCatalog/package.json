{
  "name": "@kongponents/kcardcatalog",
  "version": "6.1.16",
  "description": "A grid view of KCards",
  "main": "dist/KCardCatalog.umd.min.js",
  "componentName": "KCardCatalog",
  "source": "index.js",
  "files": [
    "dist",
    "*.vue",
    "index.js"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^6.1.16",
    "@kongponents/kcard": "^6.1.16",
    "@kongponents/kemptystate": "^6.1.16",
    "@kongponents/kpagination": "^6.1.16",
    "@kongponents/kskeleton": "^6.1.16",
    "@kongponents/utils": "^6.1.16",
    "vue": "2.6.10"
  },
  "peerDependencies": {
    "@vue/composition-api": "^1.2.4"
  }
}
