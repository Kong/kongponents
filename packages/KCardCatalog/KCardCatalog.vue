<template>
  <div class="k-card-catalog">
    <div
      v-if="title"
      class="k-card-catalog-title">
      <h2>{{ title }}</h2>
    </div>
    <div
      :class="`k-card-${cardSize}`"
      class="k-catalog k-catalog-page">
      <slot name="body">
        <KCatalogItem
          v-for="item in items"
          :key="item.key ? item.key : null"
          :item="item"
          :location-param="item.locationParam ? item.locationParam : ''"
          :truncate="!noTruncation"
          class="catalog-item"
        />
      </slot>
    </div>
  </div>
</template>

<script>
import KCatalogItem from './KCatalogItem.vue'

export default {
  name: 'KCardCatalog',
  components: {
    KCatalogItem
  },
  props: {
    items: {
      type: Array,
      default: () => []
    },
    /* pageSize: {
      type: Number,
      default: 12
    },
    totalCount: {
      type: Number,
      default: 0
    },
    searchTriggered: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    }, */
    cardSize: {
      type: String,
      default: 'medium'
    },
    title: {
      type: String,
      default: ''
    },
    noTruncation: {
      type: Boolean,
      default: false
    }
  },
  methods: {
    /* onPageChanged (page) {
      this.$emit('page-changed', page)
    } */
  }
}
</script>

<style lang="scss">
.k-card-catalog {
  $cardHeight: 181px;

  .k-skeleton-grid {
    --KSkeletonCardWidth: 25%;

    .skeleton-card-column {
      padding-right: 2rem;
    }
    .skeleton-card {
      height: $cardHeight;
    }
  }

  .k-catalog {
    display: grid;
    grid-gap: var(--spacing-lg);
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));

    &.k-card-small {
      grid-template-columns: repeat(auto-fill, minmax(150px, 1fr));
    }

    &.k-card-large {
      grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    }

    a.catalog-item, a.catalog-item:focus, a.catalog-item:hover {
      text-decoration: none;
    }
  }
}
</style>
