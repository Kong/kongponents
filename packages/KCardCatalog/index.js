import KCardCatalog from './KCardCatalog'
import KCatalogItem from './KCatalogItem'

const Plugin = {
  install (Vue) {
    Vue.component('KCardCatalog', KCardCatalog)
    Vue.component('KCatalogItem', KCatalogItem)
  }
}

export default Plugin
export { KCardCatalog, KCatalogItem }

if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(Plugin)
}
