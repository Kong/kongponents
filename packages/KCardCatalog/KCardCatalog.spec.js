import { mount, createLocalVue } from '@vue/test-utils'
import VueCompositionAPI from '@vue/composition-api'
import KCardCatalog from '@/KCardCatalog/KCardCatalog'
// import KCatalogItem from '@/KCardCatalog/KCatalogItem'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * propsData: {
 *   testMode: true
 * }
 */

const tick = async (vm, times) => {
  for (let i = 0; i < times; ++i) {
    await vm.$nextTick()
  }
}

const largeDataSet = [
  {
    title: 'Item 1',
    description: "The item's description for number"
  },
  {
    title: 'Item 2',
    description: "The item's description for number 2"
  },
  {
    title: 'Item 3',
    description: "The item's description for number 3"
  },
  {
    title: 'Item 4',
    description: "The item's description for number 4"
  },
  {
    title: 'Item 5',
    description: "The item's description for number 5"
  },
  {
    title: 'Item 6',
    description: "The item's description for number 6"
  },
  {
    title: 'Item 7',
    description: "The item's description for number 7"
  },
  {
    title: 'Item 8',
    description: "The item's description for number 8"
  },
  {
    title: 'Item 9',
    description: "The item's description for number 9"
  },
  {
    title: 'Item 10',
    description: "The item's description for number 10"
  }
]

// Use the Composition API
beforeEach(() => {
  const localVue = createLocalVue()

  localVue.use(VueCompositionAPI)
})

describe('KCardCatalog', () => {
  function getItems (count) {
    let myItems = []

    for (let i = 0; i < count; i++) {
      myItems.push({
        title: 'Item ' + i,
        description: "The item's description for number " + i
      })
    }

    return myItems
  }

  const longItem = {
    title: 'A very long item',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas in tempus lorem, et molestie quam. Praesent sapien massa, posuere et volutpat nec, imperdiet a dui. Fusce non leo posuere, molestie neque et, posuere ex. Nullam euismod tortor in est sagittis iaculis. In sodales bibendum nulla nec feugiat.'
  }

  describe('general', () => {
    it('renders proper cards when using props', () => {
      const title = 'Cool beans!'
      const wrapper = mount(KCardCatalog, {
        propsData: {
          title,
          items: getItems(5),
          testMode: true
        }
      })

      expect(wrapper.find('.k-card-catalog-title').html()).toEqual(expect.stringContaining(title))
      expect(wrapper.find('.k-catalog-page').exists()).toBeTruthy()
      // expect(wrapper.findAll('.k-card-catalog-item')).toHaveLength(5)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders slots when passed', () => {
      const slotContent = 'Look mah! No props (except testMode)'

      const wrapper = mount(KCardCatalog, {
        propsData: {
          testMode: true
        },
        scopedSlots: {
          'body': `<span>${slotContent}</span>`
        }
      })

      expect(wrapper.find('.k-catalog-page').html()).toEqual(expect.stringContaining(slotContent))
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders slots when passed (with empty)', async () => {
      const emptySlotContent = 'Look mah! I am empty! (except testMode)'

      const wrapper = mount(KCardCatalog, {
        propsData: {
          testMode: true,
          isLoading: false,
          fetcher: () => { return { data: [] } }
        },
        scopedSlots: {
          'empty-state': `<span>${emptySlotContent}</span>`
        }
      })

      await tick(wrapper.vm, 1)

      expect(wrapper.find('[data-testid="k-card-catalog-empty-state"]').html()).toEqual(expect.stringContaining(emptySlotContent))
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('renders slots when passed (with error)', () => {
      const errorSlotContent = 'Look mah! I am erroneous! (except testMode)'

      const wrapper = mount(KCardCatalog, {
        propsData: {
          testMode: true,
          hasError: true
        },
        scopedSlots: {
          'error-state': `<span>${errorSlotContent}</span>`
        }
      })

      expect(wrapper.find('[data-testid="k-card-catalog-error-state"]').html()).toEqual(expect.stringContaining(errorSlotContent))
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('can change card sizes - small', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: getItems(5),
          cardSize: 'small',
          testMode: true
        }
      })

      expect(wrapper.props('cardSize')).toBe('small')
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('can change card sizes - large', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: getItems(5),
          cardSize: 'large',
          testMode: true
        }
      })

      expect(wrapper.props('cardSize')).toBe('large')
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('handles truncation', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: [longItem],
          testMode: true
        }
      })

      expect(wrapper.props('noTruncation')).toBe(false)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('can disable truncation', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: [longItem],
          noTruncation: true,
          testMode: true
        }
      })

      expect(wrapper.props('noTruncation')).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('matches snapshot', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          testMode: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('states', () => {
    it('displays an empty state when no data is available', async () => {
      const fetcher = () => new Promise(resolve => resolve({ data: [] }))
      const wrapper = mount(KCardCatalog, {
        propsData: {
          fetcher: fetcher,
          pageSize: 4
        }
      })

      await tick(wrapper.vm, 1)

      expect(wrapper.html()).toContain('empty-state-wrapper')
      // expect(wrapper.html()).toMatchSnapshot()
    })

    it('displays a loading skeletion when the "isLoading" prop is set to true"', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: [],
          isLoading: true,
          testMode: true
        }
      })

      expect(wrapper.props('isLoading')).toBe(true)
      expect(wrapper.html()).toMatchSnapshot()
    })

    it('displays an error state when the "hasError" prop is set to true"', () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          options: [],
          hasError: true,
          testMode: true
        }
      })

      expect(wrapper.find('.empty-state-wrapper.is-error').exists()).toBeTruthy()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('pagination', () => {
    it('displays pagination when fetcher is provided', async () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          fetcher: () => {
            return largeDataSet
          },
          isLoading: false,
          testMode: true,
          paginationPageSizes: [10, 20, 30, 40]
        }
      })

      await tick(wrapper.vm, 1)

      expect(wrapper.find('[data-testid="k-pagination-container"]').exists()).toBe(true)
    })

    it('allows disabling pagination', async () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          fetcher: () => {
            return largeDataSet
          },
          isLoading: false,
          testMode: true,
          paginationPageSizes: [10, 20, 30, 40],
          disablePagination: true
        }
      })

      await tick(wrapper.vm, 1)

      expect(wrapper.find('[data-testid="k-pagination-container"]').exists()).toBe(false)
    })

    it('does not display pagination when no fetcher', async () => {
      const wrapper = mount(KCardCatalog, {
        propsData: {
          items: [],
          paginationPageSizes: [10, 20, 30, 40]
        }
      })

      await tick(wrapper.vm, 1)

      expect(wrapper.find('[data-testid="k-pagination-container"]').exists()).toBe(false)
    })
  })
})
