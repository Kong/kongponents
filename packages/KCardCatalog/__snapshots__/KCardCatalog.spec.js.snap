// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KCardCatalog general can change card sizes - large 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-large">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general can change card sizes - small 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-small">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general can disable truncation 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general handles truncation 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general matches snapshot 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general renders proper cards when using props 1`] = `
<div class="k-card-catalog">
  <div class="k-card-catalog-title">
    <h3>Cool beans!</h3>
  </div>
  <div class="k-catalog-page k-card-medium">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general renders slots when passed (with empty) 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="" data-testid="k-card-catalog-empty-state"><span>Look mah! I am empty! (except testMode)</span></div>
</div>
`;

exports[`KCardCatalog general renders slots when passed (with error) 1`] = `
<div class="k-card-catalog">
  <!---->
  <div data-testid="k-card-catalog-error-state"><span>Look mah! I am erroneous! (except testMode)</span></div>
</div>
`;

exports[`KCardCatalog general renders slots when passed 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium"><span>Look mah! No props (except testMode)</span>
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog general renders slotted cards when passed 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium">
    <section aria-labelledby="test-title-id-1234" aria-describedby="test-content-id-1234" class="kong-card grid-item d-flex flex-column overflow-hidden k-card-catalog-item catalog-item border hover" data-testid="Item-0-catalog-item" role="button" tabindex="0">
      <div class="k-card-header d-flex mb-4">
        <div>
          <!---->
          <div id="test-title-id-1234" class="k-card-title">
            <h4><span>Look mah!</span></h4>
          </div>
        </div>
        <div class="k-card-actions"></div>
      </div>
      <div class="k-card-content d-flex">
        <div id="test-content-id-1234" class="k-card-body">
          <div class="multi-line-truncate"><span>My body</span></div>
        </div>
        <!---->
      </div>
    </section>
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog states displays a loading skeletion when the "isLoading" prop is set to true" 1`] = `
<div class="k-card-catalog">
  <!---->
  <div class="k-catalog-page k-card-medium">
    <!---->
  </div>
</div>
`;

exports[`KCardCatalog states displays an error state when the "hasError" prop is set to true" 1`] = `
<div class="k-card-catalog" options="">
  <!---->
  <div data-testid="k-card-catalog-error-state">
    <section class="empty-state-wrapper is-error">
      <div class="empty-state-title">
        <div class="k-empty-state-icon card-icon mb-4 warning-icon"><span class="kong-icon kong-icon-warning"><svg width="48px" height="42px" viewBox="0 0 48 42" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <title>Warning</title>
  <g id="Page-1" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
    <g id="icn-warning-64x">
      <path d="M25.129733,2.64556484 C25.0166566,2.45204254 24.8546193,2.29069544 24.6593031,2.17754095 C24.0339615,1.81525556 23.2314423,2.02651215 22.8697255,2.64556484 L2.17517915,38.0628158 C2.06036961,38.2593042 2,38.4821557 2,38.7088951 C2,39.4200371 2.58244376,40 3.30518292,40 L44.6942755,40 C44.9262204,40 45.1538685,39.9388552 45.3538494,39.8229983 C45.9741835,39.4636139 46.1830669,38.6768553 45.8242793,38.0628158 L25.129733,2.64556484 Z M26.8565586,1.63656862 L47.5511049,37.0538196 C48.4690879,38.6248825 47.9342141,40.6394832 46.3564308,41.5535561 C45.8517121,41.8459599 45.2782053,42 44.6942755,42 L3.30518292,42 C1.4797808,42 3.66949018e-16,40.5265221 0,38.7088951 C0,38.1274525 0.154699044,37.5563885 0.448353557,37.0538196 L21.1428999,1.63656862 C22.0608829,0.0655056975 24.0841012,-0.467089888 25.6618845,0.446983085 C26.1568841,0.733756171 26.5685588,1.14367738 26.8565586,1.63656862 Z" id="Triangle" fill="#A3BBCC" fill-rule="nonzero"></path>
      <path d="M25.129733,2.64556484 C25.0166566,2.45204254 24.8546193,2.29069544 24.6593031,2.17754095 C24.0339615,1.81525556 23.2314423,2.02651215 22.8697255,2.64556484 L2.17517915,38.0628158 C2.06036961,38.2593042 2,38.4821557 2,38.7088951 C2,39.4200371 2.58244376,40 3.30518292,40 L44.6942755,40 C44.9262204,40 45.1538685,39.9388552 45.3538494,39.8229983 C45.9741835,39.4636139 46.1830669,38.6768553 45.8242793,38.0628158 L25.129733,2.64556484 Z" id="Path" fill="#FFFFFF" fill-rule="nonzero" type="secondary"></path>
      <path d="M23,13 L25,13 C25.5522847,13 26,13.4477153 26,14 C26,16 26,18 26,20 C26,22.6594889 25.7904331,25.0570193 25.3712994,27.192591 C25.2792336,27.6616739 24.8680521,28 24.3900199,28 L23.6099609,27.9999816 C23.1319364,27.9999816 22.7207615,27.6616608 22.6286994,27.1925851 C22.2095665,25.0570148 22,22.6594864 22,20 C22,18 22,16 22,14 C22,13.4477153 22.4477153,13 23,13 Z M24,31 C25.1045695,31 26,31.8954305 26,33 C26,34.1045695 25.1045695,35 24,35 C22.8954305,35 22,34.1045695 22,33 C22,31.8954305 22.8954305,31 24,31 Z" id="!" fill="#A3BBCC"></path>
    </g>
  </g>
</svg>
</span></div>
        <div class="k-empty-state-title-header">An error occurred</div>
      </div>
      <div class="empty-state-content">
        <div class="k-empty-state-message">Data cannot be displayed due to an error.</div>
        <div class="k-empty-state-cta">
          <!---->
        </div>
      </div>
    </section>
  </div>
</div>
`;
